test/vault/secrets/kvv2_test.clj

```diff
@@ -0,0 +1,133 @@
+(ns vault.secrets.kvv2-test
+  (:require
+    [clojure.test :refer [testing deftest is]]
+    [vault.client.http :as http-client]
+    [vault.secrets.kvv2 :as vault-kv])
+  (:import
+    (clojure.lang
+      ExceptionInfo)))
+
+
+(deftest write-config!-test
+  (let [mount "mount"
+        token-passed-in "fake-token"
+        vault-url "https://vault.example.amperity.com"
+        client (vault.core/new-client vault-url)
+        new-config {:max_versions 5
+                    :cas_require false
+                    :delete_version_after "3h25m19s"}]
+    (vault.core/authenticate! client :token token-passed-in)
+    (testing "Config can be updated with valid call"
+      (with-redefs
+        [clj-http.client/request
+         (fn [req]
+           (is (= :post (:method req)))
```

Comment:
I tried to make it as as agnostic as possible to the internal implementation, which is why I mocked  at the http request level, and wrote these against the Vault API as opposed to against the code. For any `write` or `delete` especially, these are arguably the most important tests (testing the structure of the request). Do you have any suggestions on other ways to test this?

We could inject the http client into the vault client, and then I could create a mock one that allows me to set up similar assertions? But that seems just as brittle

I could use something like this: https://github.com/myfreeweb/clj-http-fake
But it just use redefs under the hood like I did here, and I still would want to assert on the structure of the request which it doesn't really help with

Reply:
I guess I'm not sure why a kvv2 unit test should care specifically about the client implementation, this seems like a boundary-crossing thing. If it were an integration or functional test (and was marked with metadata as such so it could be run separately) I could see testing this end-to-end like that.

Otherwise, the unit test should be testing (externally known) state before and after a transition such as `write-config!`.

Reply:
I wouldn't really call this an integration test since I'm not actually running a vault server, but it definitely is a functional test. But I think functional testing and unit testing for publicly exposed functions sometimes overlap, as in this case. I'm also not sure it matters how it's labelled, since it's a fast test that should be run by CI after every change anyway.

I don't really think it's possible to meaningfully test a CLI that exposes an HTTP API without checking the requests being sent to the API endpoint are valid. No matter how I test it, I will have to mock the http request. So the question is whether that mock includes assertions on the data passed to it. This is not really an internal implementation detail (except for the fact that we need to know that `clj-http` is being used), since the format of the request is implementation agnostic (in fact these tests were written prior to writing the code, and the implementation has changed significantly since the tests were written without having to rewrite them)

Reply:
I see what you're saying in terms of needing to test against a known http endpoint: this should confirm it meets the contract of what that endpoint is expecting. I am probably just getting hung up on naming and semantics not so much arguing about the validity of this kind of testing.

The test says (from the `testing` form above): "Config can be updated with valid call", yet is not testing that portion of the contract or behavior since the http request returns a blank `:status 200`.

It is testing "can an authenticated http client form a correct request to write a secret to kvv2". Not trying to be prescriptive of naming here, trying to point out what these tests are actually trying to accomplish should match stated goals. I think it would sit better with me if the intent of the test is more clear in naming, description.

As I look around more, this description matching reality is generally true (eg https://github.com/amperity/vault-clj/pull/35/files/d373794059e28d008a50fef7f3cca91185e0033c#diff-be1f5f4507061b33600be2eea08d1651R85), I guess I've just narrowly noticed this, and its a fairly minor detail to fix.

> the format of the request is implementation agnostic (in fact these tests were written prior to writing the code, and the implementation has changed significantly since the tests were written without having to rewrite them)

this seems like a nice signal that these assertion then are not brittle, sounds good!

Reply:
> It is testing "can an authenticated http client form a correct request to write a secret to kvv2". Not trying to be prescriptive of naming here, trying to point out what these tests are actually trying to accomplish should match stated goals. I think it would sit better with me if the intent of the test is more clear in naming, description.

That's a good point! I'll make the naming more descriptive, the description currently makes it sound more like an integration test than it is